### PR TITLE
Add more integration test for undelete in ambry-server

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
Adding more integration tests for undelete in ambry-server
Shutdown selector in socket network client.

### CODE REVIEW

**Comment by zzmao** [comment]
Where is the code for `Shutdown selector in socket network client.`?

**Comment by justinlin-linkedin** [comment]
> Where is the code for `Shutdown selector in socket network client.`?

It's in Selector.java.  Before this PR, we didn't shutdown the executorpool in the selector. Hence, in the test, we leave lots of threads running until it reaches the limit of number of thread we can create in jvm and fails. This PR would fix it.

File: ambry-network/src/main/java/com/github/ambry/network/Selector.java [hunk 1]
```diff
@@ -243,6 +244,9 @@ public void close() {
       metrics.selectorNioCloseErrorCount.inc();
       logger.error("Exception closing nioSelector:", e);
     }
+    if (executorPool != null) {
+      executorPool.shutdown();
```
Comment (jsjtzyy) [inline ambry-network/src/main/java/com/github/ambry/network/Selector.java:248]: Good fix. Btw, could you format this file? (At least `import java.util.concurrent.Executors;` can be removed)

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java [hunk 2]
```diff
@@ -2051,6 +2059,290 @@ static void endToEndReplicationWithMultiNodeSinglePartitionTest(String routerDat
...
+      notificationSystem.awaitBlobUndeletes(blobId2.toString());
+
+      // Now after replication is resumed, the undelete of lifeversion 2 will eventually be replicated to all servers.
+      partitionRequestInfoList = getPartitionRequestInfoListFromBlobId(blobId2);
+      for (ConnectedChannel connectedChannel : channels) {
+        // Even if the notificationSystem acknowledged the undelete, it might be triggered by undelete at lifeversion 1.
+        // So check in a loop with a time out.
+        long deadline = System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(10);
+        while (true) {
+          GetRequest getRequest =
```
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java:2288]: minor: format this file

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java [hunk 3]
```diff
@@ -2051,6 +2059,290 @@ static void endToEndReplicationWithMultiNodeSinglePartitionTest(String routerDat
...
+    short blobIdVersion = CommonTestUtils.getCurrentBlobIdVersion();
+    Map<String, List<DataNodeId>> dataNodesPerDC =
+        clusterMap.getDataNodes().stream().collect(Collectors.groupingBy(d -> d.getDatacenterName()));
+    Map<String, Pair<SSLConfig, SSLSocketFactory>> sslSettingPerDC = new HashMap<>();
+    sslSettingPerDC.put("DC1", new Pair<>(clientSSLConfig1, clientSSLSocketFactory1));
+    sslSettingPerDC.put("DC2", new Pair<>(clientSSLConfig2, clientSSLSocketFactory2));
+    sslSettingPerDC.put("DC3", new Pair<>(clientSSLConfig3, clientSSLSocketFactory3));
+
+    List<PartitionId> partitionIds = clusterMap.getWritablePartitionIds(MockClusterMap.DEFAULT_PARTITION_CLASS);
+    DataNodeId dataNodeId = cluster.getOneDataNodeFromEachDatacenter(Arrays.asList("DC1")).get(0);
```
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java:2088]: Can be simplified to:
```
DataNodeId dataNodeId = dataNodesPerDC.get("DC1").get(0);
```

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java [hunk 4]
```diff
@@ -2051,6 +2059,290 @@ static void endToEndReplicationWithMultiNodeSinglePartitionTest(String routerDat
...
+      for (ConnectedChannel connectedChannel : new ConnectedChannel[]{channel1, channel2}) {
+        GetRequest getRequest =
+            new GetRequest(1, "clientId1", MessageFormatFlags.BlobProperties, partitionRequestInfoList,
+                GetOption.Include_All);
+        channel1.send(getRequest);
+        GetResponse getResponse =
+            GetResponse.readFrom(new DataInputStream(channel1.receive().getInputStream()), clusterMap);
+        assertEquals(ServerErrorCode.No_Error, getResponse.getPartitionResponseInfoList().get(0).getErrorCode());
+        try {
+          BlobProperties propertyOutput = MessageFormatRecord.deserializeBlobProperties(getResponse.getInputStream());
```
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java:2177]: Does this only test deserialization of blob properties? It seems `propertyOutput` is never used.

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java [hunk 5]
```diff
@@ -2051,6 +2059,290 @@ static void endToEndReplicationWithMultiNodeSinglePartitionTest(String routerDat
...
+          hasUndelete = getResponse.getPartitionResponseInfoList().get(0).getMessageInfoList().get(0).getLifeVersion()
+              == (short) 1;
+        } catch (MessageFormatException e) {
+          e.printStackTrace();
+          fail();
```
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java:2182]: seems to be redundant, can be removed

**Pull request merged**